### PR TITLE
[AIR] LightningTrainer Dolly V2 FSDP Fine-tuning Example

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

As a flagship example of LightningTrainer, introduce how to fine-tune LLM with FSDP.

Dataset: tiny_shakespear
Cluster: g4dn.8xlarge (head) + 15 x g4dn.4xlarge (workers)
Performance:

| | dolly-v2-3b | dolly-v2-7b |
| - | - | - |
| training time | 553s | 3024s |
| training + checkpointing time | 703s | 3583s |
| Cost | $2.96 | $16.2 |

Release Test Passed: 
- https://buildkite.com/ray-project/release-tests-pr/builds/37804#_
- https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_52gxtyl4m3i38iiespe43awybt

Rendered Doc: https://anyscale-ray--34990.com.readthedocs.build/en/34990/ray-air/examples/dolly_lightning_fsdp_finetuning.html

TODO:
- [x] Keep only the 7B example
- [x] Inference on a single T4 GPU
- [x] Convert this as a release test
- [x] Add links in toc tree and other examples. 

Screenshot 2023-05-08 at 10 39 51 AM

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "checkpoint_uri = result.checkpoint.uri\n",
+    "checkpoint_local_dir = \"/tmp/model-checkpoint\"\n",
+    "# checkpoint_uri = \"s3://yunxuanx-test/model-checkpoint/finetune-dolly-v2/LightningTrainer_18e2e_00000_0_2023-04-30_17-57-08/checkpoint_000000/\"\n",
+    "\n",
+    "os.makedirs(checkpoint_local_dir, exist_ok=True)\n",
+    "os.system(f\"aws s3 sync {checkpoint_uri} {checkpoint_local_dir}\")"
```
Comment (gjoliver) [inline doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb:825]: I see. potentially this can be done via a Checkpoint class.
do these FSDP model run differently from normal PyTorch models?

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[{'generated_text': 'This is the very place, my lord, where I was born.'}]\n",
+      "[{'generated_text': 'I am a man of a thousand lives, and I will live.'}]\n",
+      "[{'generated_text': 'Once more, my lord, I beseech you, hear me speak.'}]\n"
```
Comment (gjoliver) [inline doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb:850]: pretty nice :)

File: doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb [hunk 3]
```diff
@@ -0,0 +1,1265 @@
+{
```
Comment (gjoliver) [inline doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb:1]: 3B and 7B are that different? why do you need 2 scripts?

File: doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb [hunk 4]
```diff
@@ -0,0 +1,1265 @@
...
+    "        print(self.trainer.model)\n",
+    "        return torch.optim.AdamW(self.trainer.model.parameters(), lr=self.lr, eps=self.eps)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Configure your FSDP strategy\n",
+    "As Dolly-v2-3b is a relatively large model, it cannot be properly fit into a single commercial GPU. In this example, we use the FSDP strategy to shard model parameters across multiple workers. This allows us to avoid GPU out-of-memory issues and support a larger global batch size.\n",
```
Comment (gjoliver) [inline doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb:351]: update the comments to 7b?

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb [hunk 5]
```diff
@@ -0,0 +1,888 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ":::{tips}\n",
+    "Some tips for FSDP configutarion:\n",
+    "- `sharding_strategy`:\n",
+    "    - `ShardingStrategy.NO_SHARD`: Parameters, gradients, and optimizer states are not sharded. Similar to DDP.\n",
+    "    - `ShardingStrategy.SHARD_GRAD_OP`: Gradients and optimizer states are sharded during computation, and additionally, parameters are sharded outside computation.\n",
+    "    - `ShardingStrategy.FULL_SHARD`: Parameters, gradients, and optimizer states are sharded. It has minimal GRAM usage among the 3 options.\n",
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb:335]: Yes, they correspond to 3 zero stages. good idea

File: doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb (hunk 3 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb:1]: Actually I want to only keep the 7B one. However the 7B model doesn't fit into a single GPU, thus I can't use huggingface pipeline in the inference section. Probably I can try to compress the model.
Do you have any thoughts on that?

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb (hunk 1 above)
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb:825]: I've tried `Checkpoint.from_uri()`, but the downloading speed is extremely slow. 

Actually the checkpoint has state dict for unsharded model (fsdp wrapper prefix get trimmed). We can re-shard it later by ourselves.

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb [hunk 6]
```diff
@@ -0,0 +1,888 @@
...
+    "    def __init__(self, lr=2e-5, eps=1e-8):\n",
+    "        super().__init__()\n",
+    "        self.lr = lr\n",
+    "        self.eps = eps\n",
+    "        self.model = AutoModelForCausalLM.from_pretrained(MODEL_NAME)\n",
+    "        self.predictions = []\n",
+    "        self.references = []\n",
+    "\n",
+    "    def forward(self, batch):\n",
+    "        outputs = self.model(\n",
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb:291]: yeah, normally logits come first. But if you provide `labels=` in the forward pass, the loss comes first.

File: python/ray/train/lightning/_lightning_utils.py [hunk 7]
```diff
@@ -10,23 +10,43 @@
...
 from torch.utils.data import IterableDataset, DataLoader
 from ray.data.datastream import DataIterator
 
+_LIGHTNING_GREATER_EQUAL_2_0 = Version(pl.__version__) >= Version("2.0.0")
+_TORCH_GREATER_EQUAL_1_12 = Version(torch.__version__) >= Version("1.12.0")
+
+if _LIGHTNING_GREATER_EQUAL_2_0:
+    from pytorch_lightning.strategies import FSDPStrategy
+else:
+    from pytorch_lightning.strategies import DDPFullyShardedStrategy as FSDPStrategy
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply python/ray/train/lightning/_lightning_utils.py:26]: This version is for FSDP in Lightning 1.x. We subclassed it [here](https://github.com/ray-project/ray/blob/554c1b5adb2f3c3e74f817f29532b9e76035d489/python/ray/train/lightning/_lightning_utils.py#L54)

File: doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb (hunk 3 above)
Comment (gjoliver) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb:1]: I feel like this is where you have to use Ray to do it.
Setup up a DDP worker group like what we do for the DeepSpeed inference example. Then run 1 worker per GPU.
You can then use FSDP to load the model and do forward pass under no_grad().
May not be the best solution, but a solution?

Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-7b.ipynb:1]: Thanks Jun!  I'll try this approach. This is quite important since we have no support for distributed inference right now.

Also I found that huggingface `device_map="auto"` could work. Basically it offload params that can't fit in GPU to CPU memory. I'll probably use it for this example for now.

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb [hunk 8]
```diff
@@ -0,0 +1,888 @@
...
+    "import os\n",
+    "os.environ[\"RAY_ML_DEV\"] = \"1\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Set up ray cluster \n",
+    "In this example, we are using a ray cluster with 16 g4dn.4xlarge instances. Each instance has one Tesla T4 GPU (16GiB Memory). \n",
```
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb]: discussed offline. only keep the 7B model training example

File: doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (woshiyyya) [reply doc/source/train/examples/lightning/lightning-llm-finetuning-3b.ipynb]: Update: find a correct way to get loss:

```python
outputs = self.model(...)
loss = outputs.loss
```

File: release/air_examples/dolly_v2_lightning_fsdp_finetuning/lightning-llm-finetuning-7b.ipynb [hunk 9]
```diff
@@ -0,0 +1,1218 @@
+{
```
Comment (gjoliver) [inline release/air_examples/dolly_v2_lightning_fsdp_finetuning/lightning-llm-finetuning-7b.ipynb:1]: for release tests, you can simply symlink to your example?

File: release/air_examples/dolly_v2_lightning_fsdp_finetuning/test_myst_doc.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
+"""Convert a jupytext-compliant format in to a python script
```
Comment (gjoliver) [inline release/air_examples/dolly_v2_lightning_fsdp_finetuning/test_myst_doc.py:1]: same here

File: release/air_examples/dolly_v2_lightning_fsdp_finetuning/lightning-llm-finetuning-7b.ipynb (hunk 9 above)
Comment (woshiyyya) [reply release/air_examples/dolly_v2_lightning_fsdp_finetuning/lightning-llm-finetuning-7b.ipynb]: Done!

File: release/air_examples/dolly_v2_lightning_fsdp_finetuning/test_myst_doc.py (hunk 10 above)
Comment (woshiyyya) [reply release/air_examples/dolly_v2_lightning_fsdp_finetuning/test_myst_doc.py]: Done!

**Review by gjoliver — APPROVED** [review]

**Review by richardliaw — APPROVED** [review]
